### PR TITLE
ODC-7185: Allow admins to define pre-pinned resources using a form driven experience

### PR DESCRIPTION
**Pull request by lokanandaprabhu**
**JIRA story:**
https://issues.redhat.com/browse/ODC-7185
https://issues.redhat.com/browse/ODC-7186

**Description**:
Admin is able to use the existing form-driven experience to define the default pre-pinned resources. 
 1) If the customization is not there, then default pins will be shown as pinned resource.
 2) It updates the Console CRD when any resources are saved as default.
 3) Developer tab will be disabled if the developer perspective is disabled
4) Captured telemetry events for the cluster configuration changes

**Demo:**

https://user-images.githubusercontent.com/102503482/206423965-bf8536df-b1ca-4429-a80d-6973cc3425b3.mov

Telemetry:
Screenshot 2022-12-08 at 7 52 33 PM

**How to test:**
To test this, we need https://github.com/openshift/console-operator/pull/697 and https://github.com/openshift/console/pull/12314 these 2 PR's to be merged. 

Create a cluster with these 2 PR's using below command and use that cluster for testing

`launch openshift/console-operator#697,openshift/console#12314`

To disable the developer perspective, use below commands,

Specify perspectives to bridge with ./bin/bridge

To Enable:
`./bin/bridge -perspectives '[{ "id" : "dev", "visibility": {"state": "Enabled"}}]' 
`

To Disable:
`./bin/bridge -perspectives '[{ "id" : "dev", "visibility": {"state": "Disabled"}}]' 
`

Missing Access Review:
`./bin/bridge -perspectives '[{ "id" : "dev", "visibility": {"state": "AccessReview" , "accessReview": { "missing" : [{ "resource" : "namespaces", "verb" : "list" }]}}}]'
`

Required Access Review:
`./bin/bridge -perspectives '[{ "id" : "dev", "visibility": {"state": "AccessReview" , "accessReview": { "required" : [{ "resource" : "namespaces", "verb" : "list" }]}}}]'
`

### CODE REVIEW

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.scss [hunk 1]
```diff
@@ -0,0 +1,5 @@
+.co-resource-item__tech-dev-preview {
+    color: red;
+    margin-left: 10px;
+}
+
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.scss:5]:
```suggestion
.co-resource-item__tech-dev-preview {
    color: red;
    margin-left: 10px;
}

```

File: frontend/packages/console-app/src/actions/providers/perspective-state-provider.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    } else if (perspectives[index].visibility.state === PerspectiveVisibilityState.AccessReview) {
+      hasReviewAccess(perspectives[index]?.visibility?.accessReview)
+        .then((res) => {
+          setFeatureFlag(FLAG_DEVELOPER_PERSPECTIVE, res);
+        })
+        .catch((e) => {
+          // eslint-disable-next-line no-console
+          console.warn('Could not check access', e);
+        });
+    }
```
Comment (debsmita1) [inline frontend/packages/console-app/src/actions/providers/perspective-state-provider.ts:30]:
```suggestion
    const devPerspective = perspectives?.find((p) => p.id === 'dev');
    if (!devPerspective) {
      setFeatureFlag(FLAG_DEVELOPER_PERSPECTIVE, true);
    } else if (devPerspective.visibility.state === PerspectiveVisibilityState.Disabled) {
      setFeatureFlag(FLAG_DEVELOPER_PERSPECTIVE, false);
    } else if (devPerspective.visibility.state === PerspectiveVisibilityState.Enabled) {
      setFeatureFlag(FLAG_DEVELOPER_PERSPECTIVE, true);
    } else if (devPerspective.visibility.state === PerspectiveVisibilityState.AccessReview) {
      hasReviewAccess(devPerspective?.visibility?.accessReview)
        .then((res) => {
          setFeatureFlag(FLAG_DEVELOPER_PERSPECTIVE, res);
        })
        .catch((e) => {
          // eslint-disable-next-line no-console
          console.warn('Could not check access', e);
        });
    }
```

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+  React.useEffect(() => {
+    if (consoleConfig && consoleConfigLoaded && !configuredPerspectives) {
+      const perspectiveDetails = consoleConfig?.spec?.customization?.perspectives;
+      const index = perspectiveDetails?.findIndex((p) => p.id === 'dev');
+      let defaultPinnedResources = [];
+      if (
+        index === -1 ||
+        index === undefined ||
+        (index !== -1 && perspectiveDetails[index]?.pinnedResources === undefined)
+      ) {
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx:178]:
```suggestion
      const devPerspective = perspectiveDetails?.find((p) => p.id === 'dev');
      let defaultPinnedResources = [];
      if (
       !devPerspective ||
      !devPerspective?.pinnedResources){
```

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+              version: resource?.apiVersion,
+              resource: resource?.plural,
+            };
+          });
+        }
+        setPinnedResources(defaultPinnedResources);
+        setPinnedResourcesConfigured(defaultPinnedResources);
+      } else {
+        setPinnedResources(perspectiveDetails[index].pinnedResources);
+        setPinnedResourcesConfigured(perspectiveDetails[index].pinnedResources);
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx:200]:
```suggestion
        setPinnedResources(devPerspective.pinnedResources);
        setPinnedResourcesConfigured(devPerspective.pinnedResources);
```

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+        setSaveStatus({ status: 'successful' });
+      })
+      .catch((error) => {
+        setPinnedResources(pinnedResourcesConfigured);
+        setSaveStatus({ status: 'error', error });
+      });
+  }, 2000);
+
+  const onListChange = (
+    newEnabledOptions: React.ReactElement<ItemProps>[],
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx:277]: do we need this ?

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    const newPinnedResources = validResources?.map((resource) => {
+      return {
+        group: resource?.props?.model?.apiGroup ? resource?.props?.model?.apiGroup : '',
+        version: resource?.props?.model?.apiVersion,
+        resource: resource?.props?.model?.plural,
+      };
+    });
+    setPerspectiveData(() => {
+      const newConfiguredPerspectives = configuredPerspectives ? [...configuredPerspectives] : [];
+      const index = newConfiguredPerspectives?.findIndex((p) => p.id === 'dev');
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx:290]: same as mentioned above, use `find`

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx (hunk 5 above)
Comment (lokanandaprabhu) [reply frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx:277]: From DualListSelector of Patternfly, onListChange 2 attributes are being passed, new options and old options.. and in all other places new option is kept simply or printed in logs if it is not used... So I kept it.

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.scss [hunk 7]
```diff
@@ -0,0 +1,4 @@
+.co-resource-item__tech-dev-preview {
+    color: red;
+    margin-left: 10px;
+}
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.scss:4]: add a new line after this

File: frontend/packages/dev-console/src/components/catalog/CatalogTypesConfiguration.tsx [hunk 8]
```diff
@@ -239,6 +239,11 @@ const CatalogTypesConfiguration: React.FC<{ readonly: boolean }> = ({ readonly }
 
   return (
     <FormSection title={t('devconsole~Developer catalog')} data-test="catalog-types form-section">
+      <FormHelperText isHidden={false}>
+        {t(
+          'devconsole~Another option to customize and standardize your development process. As admin you can disable the complete Developer Catalog, or individual subcatalogs (available as Types in the Developer Catalog). Also here the "Search" and "Topology" will still show such resources.',
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/catalog/CatalogTypesConfiguration.tsx:244]:
```suggestion
          'devconsole~Another option to customize and standardize your development process. As an admin, you can disable the complete Developer Catalog, or individual sub-catalogs (available as Types in the Developer Catalog). Also here the "Search" and "Topology" will still show such resources.',
```

**Review by debsmita1 — CHANGES_REQUESTED** [review]

**Comment by debsmita1** [comment]
tested this locally
/lgtm

File: frontend/packages/dev-console/console-extensions.json [hunk 9]
```diff
@@ -56,6 +56,35 @@
       ]
     }
   },
+  {
+    "type": "console.cluster-configuration/item",
+    "properties": {
+      "id": "dev-console.PinnedResourcesConfiguration",
+      "groupId": "developer",
+      "label": "%devconsole~Pre-pinned navigation items%",
+      "description": "%devconsole~A list of default resouces saved.%",
```
Comment (jerolimov) [inline frontend/packages/dev-console/console-extensions.json:65]: typo, and then I would recommend to help the cluster-admin what does this mean exactly:

```suggestion
      "description": "%devconsole~Resource links that are shown below the standard navigation items. These pinned resources can still be changed by the user.%",
```

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+import {
+  Perspective,
+  PerspectivePinnedResource,
+  PerspectiveVisibilityState,
+  usePerspectives,
+} from '@console/shared/src/hooks/perspective-utils';
+import './PinnedResourcesConfiguration.scss';
+
+// Blacklist known duplicate resources.
+const blacklistGroups = ImmutableSet([
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx:40]: We should avoid a nomenclature like this, please use a term like `skipDuplicateGroups` or `hideGroups`.

And is this really needed at all if you only show primary CRDs?

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+        // Remove blacklisted items.
+        if (
+          blacklistGroups.has(apiGroup) ||
+          blacklistResources.has(`${apiGroup}/${apiVersion}.${kind}`)
+        ) {
+          return false;
+        }
+
+        // Only show resources that can be listed.
+        if (!_.isEmpty(verbs) && !_.includes(verbs, 'list')) {
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx:105]: Isn't this the same without lodash?

```suggestion
        if (!verbs?.some(v => v === 'list')) {
```

**Comment by jerolimov** [comment]
/retitle ODC-7185: Allow admins to define pre-pinned resources using a form driven experience
/retest

**Review by jerolimov — CHANGES_REQUESTED** [review]
See below:

File: frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx (hunk 10 above)
Comment (lokanandaprabhu) [reply frontend/packages/dev-console/src/components/catalog/PinnedResourcesConfiguration.tsx]: I changed the wordings. I followed the same logic used in Search page.

**Review by jerolimov — APPROVED** [review]
@lokanandaprabhu Thanks for the latest changes. @debsmita1 Code/direction looks fine to me, but FYI, I didn't tested it myself. Let me know if I should do this otherwise its up to you. :)

/approve

**Comment by sanketpathak** [comment]
Verified on cluster using cluster-bot
[Screencast from 12-23-2022 03:29:24 PM.webm](https://user-images.githubusercontent.com/10252230/209532011-bc51c635-fe85-4705-8bcd-fa966f251fe6.webm)
/label qe-approved

**Comment by sanketpathak** [comment]
Propagating the lgtm from above
/lgtm

**Comment by debsmita1** [comment]
Propagating docs and px acks from the Epic
/label docs-approved
/label px-approved

**Pull request merged**